### ISSUE TITLE
Discover ARP option is not being deployed

### ISSUE THREAD

**Issue by domy86** [Control-D-Inc/ctrld#118]
Hi,

If I put `discover_arp = false` in ctrld config in control panel, it is not being deployed to clients while the other options are there...

![image](https://github.com/Control-D-Inc/ctrld/assets/84593035/4f738001-9b09-4502-b4e4-464db49186e6)

**Comment by cuonglm** [comment]
Ops, thanks for reporting.

This is a bug in marshaling config, we will address this soon.

**Comment by yegors** [comment]
We will fix the bug, but may I ask why you wish to disable all discovery? This is a major feature and offers several advantages if you're using the latest version. This includes sending client data along with DNS queries, but also serving as the authoritative local DNS server, that can resolve LAN-local PTR and A records for LAN clients.

**Comment by cuonglm** [comment]
@domy86 Also note that setting `discover_dhcp = false` will disable ARP discover, too, because of this bug.

**Comment by domy86** [comment]
> We will fix the bug, but may I ask why you wish to disable all discovery? This is a major feature and offers several advantages if you're using the latest version. This includes sending client data along with DNS queries, but also serving as the authoritative local DNS server, that can resolve LAN-local PTR and A records for LAN clients.

You mean I can use reverse DNS with that discovery? Because adding every new device in control panel in order to use reverse DNS is sometimes PITA.

**Comment by yegors** [comment]
You don't need to do that. It happens automatically for all LAN devices that `ctrld` discovered. 

```
[2.7.0-RELEASE][root@pfSense.home.arpa]/root: ctrld clients list
+-----------------------------------------+--------------------------------------+-------------------+------------+
|                   IP                    |               Hostname               |        Mac        | Discovered |
+-----------------------------------------+--------------------------------------+-------------------+------------+
| 10.0.10.1                               |                                      | 00:50:56:9f:0e:84 | arp        |
| 10.0.10.209                             | pfSense                              | 00:0c:29:f5:a3:55 | arp,dhcp   |
| 10.0.10.222                             | test-virtual-machine                 |                   | mdns       |
| 10.0.10.238                             | Office-Box                           | 74:56:3c:44:eb:5e | arp,mdns   |
| 10.0.10.245                             | Test-W11                             |                   | mdns       |
| 127.0.0.1                               | pfSense                              | 00:0c:29:f5:a3:55 | dhcp       |
| ::                                      | 3511c8a4-2afd-4bf3-9137-01b2fbb76889 |                   | mdns       |
| ::1                                     | pfSense                              | 00:0c:29:f5:a3:55 | dhcp       |
| fe80::20c:29ff:fef5:a355                | pfSense                              | 00:0c:29:f5:a3:55 | dhcp       |
| fe80::1554:c4ab:cfba:189                | Office-Box                           |                   | mdns       |
| fe80::ab3f:8a1c:df6b:91b9               | Test-W11                             |                   | mdns       |
+-----------------------------------------+--------------------------------------+-------------------+------------+
[2.7.0-RELEASE][root@pfSense.home.arpa]/root: dig Test-W11

; <<>> DiG 9.18.14 <<>> Test-W11
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 39472
;; flags: qr rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;Test-W11.                      IN      A

;; ANSWER SECTION:
Test-W11.               3600    IN      A       10.0.10.245

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1) (UDP)
;; WHEN: Wed Dec 20 16:36:30 UTC 2023
;; MSG SIZE  rcvd: 42

[2.7.0-RELEASE][root@pfSense.home.arpa]/root: dig -x 10.0.10.245

; <<>> DiG 9.18.14 <<>> -x 10.0.10.245
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 17696
;; flags: qr rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;245.10.0.10.in-addr.arpa.      IN      PTR

;; ANSWER SECTION:
245.10.0.10.in-addr.arpa. 0     IN      PTR     Test-W11.

;; Query time: 1 msec
;; SERVER: 127.0.0.1#53(127.0.0.1) (UDP)
;; WHEN: Wed Dec 20 16:36:37 UTC 2023
;; MSG SIZE  rcvd: 64
```

**Comment by domy86** [comment]
Hm, not getting any hostname with discoveries enabled...

```
PS C:\Users\Domy> ctrld clients list
+-----------------+-------------+-------------------+------------+
|       IP        |  Hostname   |        Mac        | Discovered |
+-----------------+-------------+-------------------+------------+
| 10.17.25.100    |             |                   | arp        |
| 10.17.25.101    |             |                   | arp        |
| 10.17.25.103    | Domy-NB     |                   | dhcp       |
| 10.17.25.105    |             |                   | arp        |
| 25.255.255.254  |             |                   | arp        |
| 127.0.0.1       | esign-local |                   | dhcp,hosts |
| 169.254.85.21   | Domy-NB     |                   | dhcp       |
| 172.26.222.169  |             |                   | arp        |
| 192.168.193.1   |             |                   | arp        |
| 192.168.193.29  |             |                   | arp        |
| 192.168.193.93  |             |                   | arp        |
| 192.168.193.100 |             |                   | arp        |
| 192.168.193.172 |             |                   | arp        |
| 192.168.193.173 |             |                   | arp        |
| 224.0.0.22      |             |                   | arp        |
| 239.255.255.250 |             |                   | arp        |
| 255.255.255.255 |             | ff:ff:ff:ff:ff:ff | arp        |
+-----------------+-------------+-------------------+------------+
```

Here is the ctrld log...

```
{"level":"notice","time":"2023-12-20T18:15:36+01:00.586","message":"Starting service"}
{"level":"notice","time":"2023-12-20T18:15:36+01:00.676","message":"Generating controld config: C:\\Windows\\system32\\ctrld.toml"}
{"level":"debug","time":"2023-12-20T18:15:36+01:00.677","message":"waiting for ctrld listener to be ready"}
{"level":"debug","time":"2023-12-20T18:15:42+01:00.766","message":"ctrld listener is ready"}
{"level":"debug","time":"2023-12-20T18:15:42+01:00.766","message":"performing self-check"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.000","message":"self-check against \"verify.controld.com\" succeeded"}
{"level":"notice","time":"2023-12-20T18:15:43+01:00.000","message":"Service started"}
{"level":"debug","iface":"Ethernet (Dock)","time":"2023-12-20T18:15:43+01:00.041","message":"setting DNS for interface"}
{"level":"debug","iface":"Ethernet (Dock)","time":"2023-12-20T18:15:43+01:00.558","message":"setting DNS successfully"}
]"}
{"level":"info","mode":"cd","time":"2023-12-20T18:15:37+01:00.454","message":"generating ctrld config from Control-D configuration"}
{"level":"info","mode":"cd","time":"2023-12-20T18:15:37+01:00.454","message":"using defined custom config of Control-D resolver"}
{"level":"info","time":"2023-12-20T18:15:37+01:00.455","message":"writing config file to: C:\\Windows\\system32\\ctrld.toml"}
{"level":"debug","iface":"Ethernet (Dock)","time":"2023-12-20T18:15:37+01:00.495","message":"setting DNS for interface"}
{"level":"debug","iface":"Ethernet (Dock)","time":"2023-12-20T18:15:37+01:00.678","message":"setting DNS successfully"}
{"level":"debug","time":"2023-12-20T18:15:37+01:00.681","message":"resolving \"dns.controld.com\" using bootstrap DNS [\"76.76.2.0:53\" \"10.17.25.100:53\" \"25.255.255.254:53\"]"}
{"level":"debug","time":"2023-12-20T18:15:37+01:00.684","message":"bootstrap IPs: [2606:1a40::22 76.76.2.22]"}
{"level":"info","time":"2023-12-20T18:15:37+01:00.684","message":"bootstrap IPs for upstream.0: [\"2606:1a40::22\" \"76.76.2.22\"]"}
{"level":"debug","iface":"Ethernet (Dock)","time":"2023-12-20T18:15:37+01:00.746","message":"checking default route interface"}
{"level":"debug","ip":"10.17.25.103","time":"2023-12-20T18:15:37+01:00.751","message":"found ip with default route interface"}
{"level":"info","time":"2023-12-20T18:15:37+01:00.751","message":"starting DNS server on listener.0: 0.0.0.0:53"}
{"level":"debug","time":"2023-12-20T18:15:37+01:00.751","message":"start hosts file discovery"}
{"level":"debug","time":"2023-12-20T18:15:37+01:00.751","message":"start dhcp discovery"}
{"level":"debug","time":"2023-12-20T18:15:37+01:00.784","message":"sending doh request to: 76.76.2.22:443"}
{"level":"debug","time":"2023-12-20T18:15:37+01:00.824","message":"start arp discovery"}
{"level":"debug","time":"2023-12-20T18:15:37+01:00.960","message":"start ptr discovery"}
{"level":"debug","time":"2023-12-20T18:15:37+01:00.961","message":"start mdns discovery"}
{"level":"debug","time":"2023-12-20T18:15:42+01:00.765","message":"start checking DNS loop"}
{"level":"debug","time":"2023-12-20T18:15:42+01:00.765","message":"skipping external: upstream.0"}
{"level":"debug","time":"2023-12-20T18:15:42+01:00.765","message":"end checking DNS loop"}
{"level":"info","time":"2023-12-20T18:15:42+01:00.790","message":"[2df99b] QUERY: 10.17.25.103:63551 (esign-local) -> listener.0: A verify.controld.com"}
{"level":"debug","time":"2023-12-20T18:15:42+01:00.791","message":"[2df99b] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:42+01:00.791","message":"[2df99b] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:42+01:00.791","message":"[2df99b] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:42+01:00.791","message":"[2df99b] sending request header: map[X-Cd-Host:[esign-local] X-Cd-Ip:[10.17.25.103] X-Cd-Mac:[9c:2d:cd:6f:b1:a2] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:42+01:00.999","message":"[2df99b] REPLY: upstream.0 -> 10.17.25.103:63551 (esign-local): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:42+01:00.999","message":"[2df99b] received response of 71 bytes in 208.7658ms"}
{"level":"info","time":"2023-12-20T18:15:43+01:00.106","message":"[3ea2ce] QUERY: 10.17.25.103:53099 (esign-local) -> listener.0: SOA www.microsoft.com"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.106","message":"[3ea2ce] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.106","message":"[3ea2ce] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.106","message":"[3ea2ce] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.106","message":"[3ea2ce] sending request header: map[X-Cd-Host:[esign-local] X-Cd-Ip:[10.17.25.103] X-Cd-Mac:[9c:2d:cd:6f:b1:a2] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:43+01:00.165","message":"[3ea2ce] REPLY: upstream.0 -> 10.17.25.103:53099 (esign-local): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.165","message":"[3ea2ce] received response of 247 bytes in 58.8489ms"}
{"level":"info","time":"2023-12-20T18:15:43+01:00.192","message":"[410baa] QUERY: 10.17.25.103:59411 (esign-local) -> listener.0: SOA www.microsoft.com"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.192","message":"[410baa] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.192","message":"[410baa] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.192","message":"[410baa] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.192","message":"[410baa] sending request header: map[X-Cd-Host:[esign-local] X-Cd-Ip:[10.17.25.103] X-Cd-Mac:[9c:2d:cd:6f:b1:a2] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:43+01:00.254","message":"[410baa] REPLY: upstream.0 -> 10.17.25.103:59411 (esign-local): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.254","message":"[410baa] received response of 247 bytes in 62.6944ms"}
{"level":"info","time":"2023-12-20T18:15:43+01:00.342","message":"[eaa6ef] QUERY: [::1]:52766 (Domy-NB) -> listener.0: SOA www.microsoft.com"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.342","message":"[eaa6ef] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.342","message":"[eaa6ef] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.342","message":"[eaa6ef] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.342","message":"[eaa6ef] sending request header: map[X-Cd-Host:[Domy-NB] X-Cd-Ip:[::1] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:43+01:00.439","message":"[eaa6ef] REPLY: upstream.0 -> [::1]:52766 (Domy-NB): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.439","message":"[eaa6ef] received response of 247 bytes in 96.6864ms"}
{"level":"info","time":"2023-12-20T18:15:43+01:00.478","message":"[b32723] QUERY: [::1]:65471 (Domy-NB) -> listener.0: SOA www.microsoft.com"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.478","message":"[b32723] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.478","message":"[b32723] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.478","message":"[b32723] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.478","message":"[b32723] sending request header: map[X-Cd-Host:[Domy-NB] X-Cd-Ip:[::1] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:43+01:00.535","message":"[b32723] REPLY: upstream.0 -> [::1]:65471 (Domy-NB): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:43+01:00.535","message":"[b32723] received response of 247 bytes in 56.9698ms"}
{"level":"info","time":"2023-12-20T18:15:44+01:00.269","message":"[05831d] QUERY: 10.17.25.103:53099 (esign-local) -> listener.0: A dns.msftncsi.com"}
{"level":"debug","time":"2023-12-20T18:15:44+01:00.269","message":"[05831d] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:44+01:00.269","message":"[05831d] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:44+01:00.269","message":"[05831d] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:44+01:00.269","message":"[05831d] sending request header: map[X-Cd-Host:[esign-local] X-Cd-Ip:[10.17.25.103] X-Cd-Mac:[9c:2d:cd:6f:b1:a2] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:44+01:00.439","message":"[05831d] REPLY: upstream.0 -> 10.17.25.103:53099 (esign-local): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:44+01:00.439","message":"[05831d] received response of 73 bytes in 170.2972ms"}
{"level":"info","time":"2023-12-20T18:15:45+01:00.445","message":"[859149] QUERY: 10.17.25.103:53099 (esign-local) -> listener.0: A eu-mobile.events.data.microsoft.com"}
{"level":"debug","time":"2023-12-20T18:15:45+01:00.445","message":"[859149] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:45+01:00.445","message":"[859149] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:45+01:00.445","message":"[859149] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:45+01:00.445","message":"[859149] sending request header: map[X-Cd-Host:[esign-local] X-Cd-Ip:[10.17.25.103] X-Cd-Mac:[9c:2d:cd:6f:b1:a2] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:45+01:00.447","message":"[fe49f6] QUERY: 10.17.25.103:51036 (esign-local) -> listener.0: AAAA eu-mobile.events.data.microsoft.com"}
{"level":"debug","time":"2023-12-20T18:15:45+01:00.447","message":"[fe49f6] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:45+01:00.447","message":"[fe49f6] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:45+01:00.447","message":"[fe49f6] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:45+01:00.448","message":"[fe49f6] sending request header: map[X-Cd-Host:[esign-local] X-Cd-Ip:[10.17.25.103] X-Cd-Mac:[9c:2d:cd:6f:b1:a2] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:45+01:00.496","message":"[859149] REPLY: upstream.0 -> 10.17.25.103:53099 (esign-local): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:45+01:00.497","message":"[859149] received response of 69 bytes in 51.5077ms"}
{"level":"info","time":"2023-12-20T18:15:45+01:00.519","message":"[fe49f6] REPLY: upstream.0 -> 10.17.25.103:51036 (esign-local): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:45+01:00.519","message":"[fe49f6] received response of 81 bytes in 71.6919ms"}
{"level":"info","time":"2023-12-20T18:15:46+01:00.482","message":"[82cab2] QUERY: 10.17.25.103:53968 (esign-local) -> listener.0: HTTPS dns.controld.com"}
{"level":"debug","time":"2023-12-20T18:15:46+01:00.482","message":"[82cab2] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:46+01:00.482","message":"[82cab2] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:46+01:00.482","message":"[82cab2] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:46+01:00.482","message":"[82cab2] sending request header: map[X-Cd-Host:[esign-local] X-Cd-Ip:[10.17.25.103] X-Cd-Mac:[9c:2d:cd:6f:b1:a2] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:46+01:00.483","message":"[178203] QUERY: 10.17.25.103:56349 (esign-local) -> listener.0: A dns.controld.com"}
{"level":"debug","time":"2023-12-20T18:15:46+01:00.484","message":"[178203] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:46+01:00.484","message":"[178203] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:46+01:00.484","message":"[178203] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:46+01:00.484","message":"[178203] sending request header: map[X-Cd-Host:[esign-local] X-Cd-Ip:[10.17.25.103] X-Cd-Mac:[9c:2d:cd:6f:b1:a2] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:46+01:00.564","message":"[82cab2] REPLY: upstream.0 -> 10.17.25.103:53968 (esign-local): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:46+01:00.564","message":"[82cab2] received response of 93 bytes in 82.3735ms"}
{"level":"info","time":"2023-12-20T18:15:46+01:00.565","message":"[178203] REPLY: upstream.0 -> 10.17.25.103:56349 (esign-local): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:46+01:00.565","message":"[178203] received response of 73 bytes in 81.5546ms"}
{"level":"info","time":"2023-12-20T18:15:50+01:00.585","message":"[786075] QUERY: 10.17.25.103:53099 (esign-local) -> listener.0: A eu-mobile.events.data.microsoft.com"}
{"level":"debug","time":"2023-12-20T18:15:50+01:00.585","message":"[786075] no explicit policy matched, using default routing -> [upstream.0]"}
{"level":"debug","time":"2023-12-20T18:15:50+01:00.585","message":"[786075] including client info with the request"}
{"level":"debug","time":"2023-12-20T18:15:50+01:00.585","message":"[786075] sending query to upstream.0: "}
{"level":"debug","time":"2023-12-20T18:15:50+01:00.585","message":"[786075] sending request header: map[X-Cd-Host:[esign-local] X-Cd-Ip:[10.17.25.103] X-Cd-Mac:[9c:2d:cd:6f:b1:a2] X-Cd-Os:[1-1-6.2]]"}
{"level":"info","time":"2023-12-20T18:15:50+01:00.646","message":"[786075] REPLY: upstream.0 -> 10.17.25.103:53099 (esign-local): NOERROR"}
{"level":"debug","time":"2023-12-20T18:15:50+01:00.646","message":"[786075] received response of 69 bytes in 60.1503ms"}
```

And I see that is listening on port 5353 but firewall shows no traffic at port 5353...

```
PS C:\Users\Domy> Get-NetUDPEndpoint -LocalPort 5353 | Select-Object LocalAddress,LocalPort,OwningProcess,@{ Name="ProcessName"; Expression={((Get-Process -Id $_.OwningProcess).Name )} }

LocalAddress   LocalPort OwningProcess ProcessName
------------   --------- ------------- -----------
::1                 5353          6164 TeamViewer_Service
::                  5353         12372 brave
169.254.214.77      5353          6164 TeamViewer_Service
0.0.0.0             5353         64380 ctrld
```

**Comment by yegors** [comment]
Initially reported bug in this thread is fixed in v1.3.4
